### DISCUSSION TITLE
Feature request: Topological sort

### DISCUSSION THREAD

**Discussion by opatut** [networkit/networkit#829] [Ideas]
I might be struggling with the docs, or there is just no toposort algorithm available in the library? I thought that would be a pretty standard graph operation everybody needs, so could it be added? Right now I've implemented my own in Python, but I bet it could be faster ;)

```python
from collections import deque

def topological_sort(G):
    queue = deque()
    in_degrees = {}

    for n in G.iterNodes():
        d = G.degreeOut(n)
        if d > 0:
            in_degrees[n] = d
        elif d == 0:
            queue.append(n)

    while queue:
        node = queue.popleft()
        yield node

        for child in G.iterInNeighbors(node):
            in_degrees[child] -= 1

            if in_degrees[child] == 0:
                queue.append(child)
                del in_degrees[child]

    if in_degrees:
        raise ValueError("Graph contains a cycle")
```

**Comment by fabratu** [comment]
Hi,

thanks for the suggestion. For your info: I am converting this issue to a discussion (we have a category for suggesting new features).

Concerning Topo Sort: We currently use Topological-Sort as part of more complex algorithms, but it is not implemented directly. However, you can use the traversal-module (`DFS` + a `visited`-tracker) to achieve the same. Since it uses the underlying C++ core, there is likely a speed-up.

Example snippet:

```
import networkit as nk

n = 2000
p = 0.15
directed = True

G = nk.generators.ErdosRenyiGenerator(n, p, directed).generate()

def topo_sort():
    visited = [False for n in G.iterNodes()]
    result = []

    def exploreNode(u):
        visited[u] = True
        result.append(u)

    for n in G.iterNodes():
        
        if(visited[n] == False):
            nk.graph.Traversal.DFSfrom(G, n, exploreNode)

    print(len(result))
```

Cycle detection like [here](https://en.wikipedia.org/wiki/Topological_sorting) has also to be added in the end.

**Reply by opatut** [reply]
_Edit: I'm wrong, see comment below._

Thank you for your ideas. Sadly, DFS does not fulflil the needs here, it will only work correctly on graphs that don't branch and reunite, e.g. 

```
A -> B -> C -> D
A -> B -> E -> D
```

DFS from A will visit in order A, B, C, D, E, but this breaks the E -> D edge.

BFS has a similar problem on this tree:

```
A -> B -> C -> D
A -> B -> E -> F -> D
```

It will visit in order A, B, C, E, D, F and break the F -> D edge.

Sorry I'm not good with the mathematical names for these things, I hope you understand my issue. AFAIK there is no way around tracking the frontier of nodes with zero indegree to implement a toposort...

**Reply by opatut** [reply]
Oh wow nevermind, I just saw which part of the Wikipedia page you're refering to. Yeah well, with prepending to the output, that makes a lot more sense now. Indeed, that could work.

Will we get a public interface for doing this though, getting rid of the headaches for normies like me? :pray: :blush:

**Reply by fabratu** [reply]
Yes, the prepending the node and visiting all neighbors first makes topological sort possible with DFS 👍  You are right concerning BFS though.

> Will we get a public interface for doing this though, getting rid of the headaches for normies like me? 🙏 😊

Cannot promise it, but since it is a common functionality we will consider it of course 🙂 You (or anyone else who is interested) can also contribute a patch at anytime.

**Comment by avdgrinten** [comment]
IMO, this is a reasonable request. Even if DFS can be used for this purpose, a full C++ implementation is likely much faster.

**Comment by fabratu** [comment]
Currently we have an open PR #840.  When it is merged and you want to use the algorithm/interface right away, you can download the wheel-artifacts from the CI-run (https://github.com/networkit/networkit/actions/workflows/build_wheels.yml). Best wait for a finished build on the branch `master`.

**Comment by meffertj** [comment]
this discussion should get closed, since topsort is available now, no?
